### PR TITLE
Neurify

### PR DESCRIPTION
**Pull request by denismazzucato**
Hi Caterina, here is the PR for Neurify domain. 

I may have added something you do not want in the master branch (as the forward runner forward_runner.py or the testing script for neurify neurify_test.py). Take a look and let me know what to change!

### CODE REVIEW

File: src/libra/engine/bias_analysis.py [hunk 1]
```diff
@@ -32,12 +33,14 @@
 from libra.engine.runner import Runner
 from libra.frontend.cfg_generator import ast_to_cfg
 
+SYMB_DOMAINS_TUPLE = (DeepPolyState, NeurifyState)
```
Comment (caterinaurban) [inline src/libra/engine/bias_analysis.py]: I would rename this 'NON_APRON_DOMAINS'

File: src/libra/engine/backward.py [hunk 2]
```diff
@@ -137,7 +138,7 @@ def feasibility(self, state, manager, key=None, do=False, chunk=None):
         outcomes = set()
         _disjunctions = None
         for idx, value in enumerate(self.values):
-            if isinstance(state, DeepPolyState):
+            if isinstance(state, libra.engine.bias_analysis.SYMB_DOMAINS_TUPLE):
```
Comment (caterinaurban) [inline src/libra/engine/backward.py]: Why don't you import the constant? (Same for all instances below)

File: src/libra/engine/forward.py [hunk 3]
```diff
@@ -1,84 +1,125 @@
...
+        """Forward control flow graph interpreter construction.
+
+        :param cfg: control flow graph to analyze
+        :param semantics: semantics of statements in the control flow graph
+        :param widening: number of iterations before widening
+        :param precursory: precursory control flow graph interpreter
+        """
+        super().__init__(cfg, semantics, widening)
+        self.manager = manager
+        self._avoid_state_log = False
```
Comment (caterinaurban) [inline src/libra/engine/forward.py]: Mh. Too many negations. Can we just have `self._state_log` or even just `self._log`?

File: src/libra/engine/forward_runner.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+from libra.abstract_domains.deeppoly_domain import DeepPolyState
+from libra.abstract_domains.neurify_domain import NeurifyState
+from libra.abstract_domains.interval_domain import BoxState
+from libra.abstract_domains.symbolic1_domain import Symbolic1State
+from libra.abstract_domains.symbolic2_domain import Symbolic2State
+from libra.core.statements import Assignment, Lyra2APRON
+from libra.frontend.cfg_generator import ast_to_cfg
+from libra.engine.forward import ForwardInterpreter, ActivationPatternForwardSemantics
+
+class ForwardRunner(Runner):
```
Comment (caterinaurban) [inline src/libra/engine/forward_runner.py]: This does not quite belong to Libra but we can keep it. For consistency with the other runner, I would rather rename it `ForwardAnalysis`, though.

**Review by caterinaurban — COMMENTED** [review]
Looks good, thanks! I have added just a few minor change requests.

**Comment by denismazzucato** [comment]
I solved the change requests and the merge conflicts, so the PR is ready to be merged! 

Note that, I solved the request about NON_APRON_DOMAINS moving out the constant definition to 'libra/abstract_domains/constants.py' since an ImportError occurred trying to import NON_APRON_DOMAINS directly from bias_analysis.py (I guess a circular dependency raised).
Tell me if this workaround suits you. If so, the PR Is ready.

More in details, I tried the following statement: 
"from libra.engine.bias_analysis import NON_APRON_DOMAINS"
Hence:
"ImportError: cannot import name 'NON_APRON_DOMAINS' from 'libra.engine.bias_analysis'"

**Pull request merged**